Allow setting userscripts directly without the need of files

Legend-Master opened the pull request:
<!-- Please describe your changes on the following line: -->

- Remove userscripts from all places and move it to servoshell
- Introduce `UserContentManager` struct and passing it through `Servo::new` all the way down to script thread instead of using `Opts::userscripts`

> https://github.com/servo/servo/pull/35388#issuecomment-2749160727

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Legend-Master committed:
Allow settings userscripts through preferences

Signed-off-by: Tony <legendmastertony@gmail.com>

Legend-Master committed:
Merge branch 'main' of https://github.com/servo/servo into userscripts-prefs

Legend-Master committed:
mach fmt instead of cargo fmt

Signed-off-by: Tony <legendmastertony@gmail.com>

Legend-Master committed:
Fix pref loading not working for array values

Signed-off-by: Tony <legendmastertony@gmail.com>

mrobinson commented on components/script/dom/window.rs:
```diff
@@ -369,10 +370,12 @@ pub(crate) struct Window {
-    /// the resources/user-agent-js directory, and if the option isn't passed userscripts
-    /// won't be loaded.
-    userscripts_path: Option<String>,
+    /// Where to load userscripts from, if any.
+    /// and if the option isn't passed userscripts won't be loaded.
+    userscripts_directory: Option<String>,
+
+    /// Inlined userscripts to run
+    userscripts: Vec<String>,
```
Please don't add any more fields to `Window` for options and settings. These aren't necessary any longer now that script depends on `servo_pref`. Instead have the code in `userscripts.rs` read the options directly.

mrobinson commented on components/script/dom/window.rs:
```diff
@@ -611,8 +614,26 @@ impl Window {
 
-    pub(crate) fn get_userscripts_path(&self) -> Option<String> {
-        self.userscripts_path.clone()
+    /// Reads all the userscript from [`Self::userscripts_directory`]
+    /// and join them with [`Self::userscripts`]
+    pub(crate) fn get_userscripts(&self) -> Vec<(String, Option<PathBuf>)> {
+        let mut userscripts: Vec<_> = self
+            .userscripts
+            .iter()
+            .map(|script| (script.clone(), None))
+            .collect();
+        if let Some(userscripts_directory) = &self.userscripts_directory {
+            let mut files = std::fs::read_dir(userscripts_directory)
+                .expect("Bad path passed to --userscripts")
+                .filter_map(|e| e.ok())
+                .map(|e| e.path())
+                .collect::<Vec<_>>();
+            files.sort();
+            for file in files {
+                userscripts.push((std::fs::read_to_string(&file).unwrap(), Some(file)));
+            }
+        }
+        userscripts
```
Please move this to `userscripts.rs`.

mrobinson reviewed (dismissed):
Can you please give me some background on why it is interesting to load userscripts via a preference? I feel like exposing this as both a preference and a option is a bit confusing.

Legend-Master commented:
> Can you please give me some background on why it is interesting to load userscripts via a preference?

I'm currently working on a Tauri integration with Servo/[Verso](https://github.com/versotile-org/verso), and Tauri has an API for running initialization scripts which it heavily depends on, and having Servo take the script strings would be much better to use than having to save the script files and pass in the directory to Servo and clean up on exit

> I feel like exposing this as both a preference and a option is a bit confusing.

Yeah, I'm not entirely sure which place to add it to be honest, I think the idea is to eventually merge `Opts` and `Preferences`? Maybe we can rename `Opts::userscripts` to `Opts::userscript_directory`?

Legend-Master committed:
Use pref! in userscripts instead

Signed-off-by: Tony <legendmastertony@gmail.com>

Legend-Master commented:
> Why not move the user script directory getter to the `EmbedderMethods` trait?

Hmm, I'm not sure how we organize these things, `EmbedderMethods` seems to only have very few things in it that are being used in `Servo::new`, I feel like it's the same as putting it in `Preferences` or else we'll just be passing it through the params all the way down instead of using it through `pref!` (or we might just be setting it to `Preferences` anyways)

Legend-Master committed:
Merge branch 'main' into userscripts-prefs

mrobinson commented:
I took a look at the[ WebKitGTK API for userscripts](https://webkitgtk.org/reference/webkitgtk/stable/method.UserContentManager.add_script.html) and I think we should work to something like that. Essentially scripts are added via strings and it is up to the embedder to load them. This should all be handled by servoshell. It shouldn't be a preference, but a method that is called on a `WebView`.

mrobinson commented:
> Something like `EmbedderMsg::GetUserscriptsToLoad(WebViewId, IpcSender<Vec<String>)`?

I think an API where the embedder can actually add and remove userscripts at any time would be much more useful as it would provide the embedder more flexibility about when and how userscripts are applied. In addition, it's very wasteful to send that much data over IPC every time we need to query the userscripts. For this API I think we should simply emulate WebKitGTK.

Legend-Master commented:
> What we can do is to have every WebView, for now, return the same UserContentManager. Later, when we have support for modifying this per-WebView we can expose the ability to override the UserContentManager per-WebView

Sorry I'm not entirely sure what you mean, I get the API you would like to expose, but I'm not sure how you would like the data inside `UserContentManager` to be passed to the script thread, like we don't have the ability to create webviews with settings right now, and also you don't like the idea of asking embedder for the userscripts every time it needs to know it because of the cost of IPC

Do you mean we make a new global `UserContentManager` and let the constellation use and pass the data inside it to script thread through `InitialPipelineState` for now or something else?

Legend-Master commented:
@mrobinson Hi, it's been a while now, do you mind considering the question again? It's a really important feature for the Tauri integration

Legend-Master committed:
Merge branch 'main' of https://github.com/servo/servo into userscripts-prefs

Legend-Master committed:
Merge branch 'main' of https://github.com/servo/servo into userscripts-prefs

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Legend-Master committed:
Merge branch 'main' into userscripts-prefs

Signed-off-by: Tony <68118705+Legend-Master@users.noreply.github.com>

jdm commented:
> Do you mean we make a new global `UserContentManager` and let the constellation use and pass the data inside it to script thread through `InitialPipelineState` for now or something else?

I think this is reasonable. InitialPipelineState is the way that we pass data from the constellation to new script threads, so I don't see another option besides sending a vector of JS sources as part of this structure.

Legend-Master committed:
Merge branch 'main' of https://github.com/servo/servo into userscripts-prefs

Legend-Master committed:
Merge branch 'userscripts-prefs' of https://github.com/Legend-Master/servo into userscripts-prefs

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Legend-Master commented:
> I think this is reasonable. InitialPipelineState is the way that we pass data from the constellation to new script threads, so I don't see another option besides sending a vector of JS sources as part of this structure.

That's good to know

So the remaining question is since we need the scripts at the very beginning and we can't create webview with settings right now, that means we either add that, or we make `UserContentManager` a global struct similar to `Opts` and `Preferences` and the user would set it like how they would set it through `Opts` or `Preferences`

In the second case, I really don't feel like there're any differences between this and just putting it in `Preferences`, and for the first one to happen, I think we'll need a lot more designs to discuss about

> So I'm thinking, maybe we could just send a message to the embedder and let the embedder sends back all the userscripts it wants to use, does that sound good to you?
> 
> Something like `EmbedderMsg::GetUserscriptsToLoad(WebViewId, IpcSender<Vec<String>)`?

Or we could do this (maybe with `IpcSharedMemory`)

jdm commented:
I see what you mean. I'm going to try to propose a set of concrete steps tonight or tomorrow to allow this PR to move forward.

Legend-Master commented:
That sounds good to me, will try to do that, marking the PR as draft for now

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Legend-Master committed:
Implement the model jdm suggested
- Remove userscripts from all places and move it to servoshell
- Add in `UserContentManager` struct and passing it through `Servo::new`
all the way down to script thread

Signed-off-by: Tony <legendmastertony@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

jdm commented on components/script/script_thread.rs:
```diff
@@ -173,7 +174,7 @@ pub(crate) fn with_script_thread<R: Default>(f: impl FnOnce(&ScriptThread) -> R)
     with_script_thread(|script_thread| {
         trace!("tracing fields of ScriptThread");
-        script_thread.trace(tr);
+        unsafe { script_thread.trace(tr) };
```
Why is this change necessary?

Legend-Master replied:
Saw an warning `call to unsafe function is unsafe and requires an unsafe function or block` so I applied the suggestion from rust analyzer, I can remove it if that's not desired

jdm replied:
I think you must be using a different version of rustc than Servo or something. Let's leave this change out of the PR.

Legend-Master replied:
Hmm that's weird, I'm using `rustc 1.85.0 (4d91de4e4 2025-02-17)`, the same as the setting in `rust-toolchain.toml`, anyways, removed

jdm commented on components/shared/embedder/user_content_manager.rs:
```diff
@@ -0,0 +1,51 @@
+use std::path::PathBuf;
```
Missing license header.

jdm commented on components/shared/embedder/user_content_manager.rs:
```diff
@@ -0,0 +1,51 @@
+        self.user_scripts.push(script.into());
+    }
+
+    pub fn get_scripts(&self) -> &Vec<UserScript> {
```
```suggestion
    pub fn scripts(&self) -> &[UserScript] {
```

jdm commented on components/script/dom/window.rs:
```diff
@@ -624,8 +624,8 @@ impl Window {
 
-    pub(crate) fn get_userscripts_path(&self) -> Option<String> {
-        self.userscripts_path.clone()
+    pub(crate) fn get_userscripts(&self) -> &Vec<UserScript> {
```
```suggestion
    pub(crate) fn userscripts(&self) -> &[UserScript] {
```

jdm reviewed (approved):
Looks good! Just a couple small points.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

jdm commented:
You'll need to ensure your commits are all signed, per bullet 3 from https://book.servo.org/contributing.html#pull-request-checklist .

Legend-Master committed:
Apply suggestions from code review and format

Signed-off-by: Tony <legendmastertony@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Legend-Master commented:
Oops, always forgot to commit signed off 😂, should be fixed now

Legend-Master committed:
Revert unrelated change

Signed-off-by: Tony <legendmastertony@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.
